src/btree/bt_npos.c

```diff
@@ -0,0 +1,362 @@
+ *     - If WT_READ_CACHE is set, the page should be in memory.
+ *     - If the initial ref is to a good page, it will be returned.
+ *     - If the initial ref is null, it does nothing.
+ */
+static int
+__find_closest_leaf(WT_SESSION_IMPL *session, WT_REF **refp, uint32_t flags)
+{
+    if (*refp == NULL || F_ISSET(*refp, WT_REF_FLAG_LEAF))
+        return (0);
+    LF_SET(WT_READ_SKIP_INTL);
+    return (__wt_tree_walk(session, refp, flags));
+}
+
+/*
+ * __page_from_npos_internal --
+ *     Go to a leaf page given its normalized position. Note that this function can return a "bad"
+ *     page (deleted, locked, etc). The caller of this function should walk the tree to find a
+ *     suitable page.
+ *
+ * NOTE: Must be called within WT_WITH_PAGE_INDEX or WT_ENTER_PAGE_INDEX
+ */
+static int
+__page_from_npos_internal(WT_SESSION_IMPL *session, WT_REF **refp, double npos, uint32_t flags)
+{
+    WT_BTREE *btree;
+    WT_DECL_RET;
+    WT_PAGE *page;
+    WT_PAGE_INDEX *pindex;
+    WT_REF *current, *descent;
+    double npos_local;
+    int idx, entries;
+    bool read_cache;
+
+    *refp = NULL;
+
+    btree = S2BT(session);
+    current = NULL;
+    /*
+     * This function is called by eviction to find a page in the cache. That case is indicated by
+     * the WT_READ_CACHE flag. Ordinary lookups in a tree will read pages into cache as needed.
+     */
+    read_cache = LF_ISSET(WT_READ_CACHE);
+
+restart: /* Restart the search from the root. */
+
+    /* Search the internal pages of the tree. */
+    current = &btree->root;
+    npos_local = npos;
+    for (;;) {
+        /*
+         * This function will always return a leaf page even if the saved position was for an
+         * internal one. This potentially can lead to internal pages being skipped by eviction in
+         * case when eviction happened to pause on an internal page and some subsequent internal
+         * pages don't have any leafs and also are subject for eviction.
+         *
+         * This is a rare case. However, it doesn't lead to completely non-evictable internal pages
+         * because eviction will eventually reach these pages during other passes.
+         */
+        if (F_ISSET(current, WT_REF_FLAG_LEAF))
+            goto done;
+
+        /* The entire for loop's body is for INTERNAL pages only */
+
+        page = current->page;
+        WT_INTL_INDEX_GET(session, page, pindex);
+        entries = (int)pindex->entries;
+
+        npos_local *= entries;
+        idx = (int)npos_local;
+        idx = WT_CLAMP(idx, 0, entries - 1);
+        npos_local -= idx;
+        descent = pindex->index[idx];
+
+        if (read_cache) {
+            /*
+             * In case of eviction, we never want to load pages from disk. Also, page_swap with
+             * WT_READ_CACHE will fail anyway and we'll lose our pointer, so avoid making a call
+             * that will fail.
+             */
+            switch (WT_REF_GET_STATE(descent)) {
```

Comment:
If I am understanding this code correctly we are currently using a parent page's index to find a ref, we then get that ref's state and if it's DISK / LOCKED / DELETED return it and pass it to `__wt_tree_walk` assuming it is not a leaf ref. 

This then splits into two cases:
1. We have a leaf ref
2. We have an internal ref

In both cases they could still be WT_REF_LOCKED. Following case 2, we call into `__find_closest_leaf` which calls `__wt_tree_walk`. That function doesn't handle being called with a locked ref, and immediately deference ref->home accessing a field on the locked ref. I think that is illegal. Similarly for a leaf ref we access `ref->home`.

Reply:
If the page is locked, then the loop doesn't descent to it.

Reply:
Makes sense, this comment: "                /* Can't go down from here but it's ok to return this page. */" made it seem like we were returning the descent page, or at least that is how I read it.

Reply:
I changed the wording.